beep/structure/base.py

```diff
@@ -0,0 +1,1373 @@
+    BEEPDatapath is an abstract base class requiring a child class to implement the following methods:
+    - from_file: Take raw cycler output files and convert them to a BEEPDatapath object. It should
+        return a BEEPDatapath object with the correct data types (specified in __init__).
+
+
+    Attributes:
+
+        Important/Very common attributes:
+            Each BEEPDatapath will have a maximum of 5 *very important* structured attributes dataframes:
+            - structured_summary (pd.DataFrame): A summary of the cycles
+            - structured_data (pd.DataFrame): The interpolated cycles
+            - diagnostic_data (pd.DataFrame): The interpolated diagnostic cycles
+            - diagnostic_summary (pd.DataFrame): A summary of diagnostic cycles
+            - raw_data (pd.DataFrame): All raw data from the raw input file.
+
+        Less important attributes:
+            - metadata (BEEPDatapath.CyclerRunMetadata): An object holding all metadata.
+            - paths (dict): A mapping of {descriptor: full_path or [paths]} for all files related to this datapath.
+                This can include things like "raw", "metadata", "structured", as well as other paths (e.g., "eis").
+
+        Private:
+            - _is_legacy (bool): Whether this file is loaded from a legacy ProcessedCyclerRun. Some operations are
+                not supported for legacy-loaded processed/structured files.
+            - _aggregation (dict): Specifies the pandas aggregation columns order for normal cycler data
+            - _diag_aggregation (dict): Specifies the pandas aggregation columns order for diagnostic data
+            - _summary_cols (list): The ordering of summary columns for normal cycler data.
+            - _diag_summary_cols (list): The ordering of summary columns for diagnostic data.
+
+    """
+
+    IMPUTABLE_COLUMNS = ["temperature", "internal_resistance"]
+
+    class StructuringDecorators:
+        """
+        Internal container class for decorators related to data structuring.
+        """
+
+        @classmethod
+        def must_be_structured(cls, func):
+            """
+            Decorator to check if the datapath has been structured.
+
+            Args:
+                func: A function or method.
+            Returns:
+                A wrapper function for the input function/method.
+            """
+
+            def wrapper(*args, **kwargs):
+                if not args[0].is_structured:
+                    raise RuntimeError(
+                        f"{args[0].__class__.__name__} has not been structured! Run .structure(*args)."
+                    )
+                else:
+                    return func(*args, **kwargs)
+
+            return wrapper
+
+        @classmethod
+        def must_not_be_legacy(cls, func):
+            """
+            Decorator to check if datapath has been serialized from legacy,
+            as some operations depend on data which is not in (or not is not easily accessed from)
```

Comment:
typo in the docstring `(or not is not easily accessed from)`.

Reply:
addressed in [9a13ca2](https://github.com/TRI-AMDD/beep/pull/150/commits/9a13ca2f1710a08355a0229a50ae21ec03e1b278)